aws-athena/views/reporting-vw_assessment_roll_muni.sql

```diff
@@ -0,0 +1,114 @@
+-- Gathers AVs by year, major class, assessment stage, and municipality for
+-- reporting
+
+/* Ensure every municipality/class/year has a row for every stage through
+cross-joining. This is to make sure that combinations that do not yet
+exist in iasworld.asmt_all for the current year will exist in the view, but have
+largely empty columns. For example: even if no class 4s in the City of Chicago
+have been mailed yet for the current assessment year, we would still like an
+empty City of Chicago/class 4 row to exist for the mailed stage. */
+WITH stages AS (
+
+    SELECT 'mailed' AS stage
+    UNION
+    SELECT 'assessor certified' AS stage
+    UNION
+    SELECT 'bor certified' AS stage
+
+),
+
+/* This CTE removes historical PINs in reporting.vw_pin_township_class that are
+not in reporting.vw_pin_value_long. We do this to make sure the samples we
+derive the numerator and denominator from for stage_portion are the same. These
+differences are inherent in the tables that feed these views (iasworld.pardat
+and iasworld.asmt_all, respectively) and are data errors - not emblematic of
+what portion of a municipality has actually progressed through an assessment
+stage.
+
+It does NOT remove PINs from the most recent year of
+reporting.vw_pin_township_class since we expect these differences based on how
+iasworld.asmt_all is populated through the year as the assessment cycle
+progresses.
+
+Starting in 2020 a small number of PINs are present in iasworld.asmt_all for
+one or two but not all three stages of assessment when we would expect all three
+stages to be present for said PINs. This is also a data error, but is NOT
+addressed in this view and leads to a few instances where stage_portion ends up
+being less than 1 when it should equal 1. 16-07-219-029-1032 missing a mailed
+value but having CCAO and BOR certified values in 2021 is an example. */
```

Comment:
🙃 Gah. Probably not worth making a view-level fix for this if it's only a few PINs right? Perhaps we could add some dbt tests? "Each PIN in the prior tax year has three stage/procname values"

Reply:
I've reached out to Mirella. dbt test is a good idea. If we were to try to address it in the view, it'd be very tricky dynamically identifying which pins are allowed to have less than three rows and which aren't.